Cancel workflow invocations when histories are deleted.

jmchilton opened the pull request:
With test case!

xref #3555 (third point)

jmchilton committed:
Cancel workflow invocations when histories are deleted.

With test case, rebased with spelling fixes thanks to @nsoranzo.

xref #3555 (third point)

nsoranzo commented on lib/galaxy/workflow/run.py:
```diff
@@ -156,6 +156,10 @@ def invoke(self):
             return self.progress.outputs
 
+        if workflow_invocation.history.deleted:
+            log.info("Cacelled workflow evaluation due to delted history")
```
s/Cacelled/Cancelled/ 
s/delted/deleted/

nsoranzo commented on test/api/test_workflows.py:
```diff
@@ -964,6 +964,31 @@ def test_workflow_request(self):
+        # Wait for all the datasets to complete, make sure the workflow invocation
+        # is not complete.
+        invocation = self._invocation_details(uploaded_workflow_id, invocation_id)
+        assert invocation['state'] != 'scheduled', invocation
```
Maybe it would be helpful to use `invocation['state'] in LIST_OF_POSSIBLE_STATES` instead, because `'failed'` would pass the assertion.

jmchilton replied:
If ``failed`` passed the assertion, the ``cancelled`` check below would fail the test. If I make it a list of ``[new, ready]`` I would need to update the list if we add more non-terminal states (one could imagine adding ``queued`` or ``waiting`` states for instance). I could reuse stuff in galaxy.models to deal with state enumerations - but I'd ultimately like the API tests to not use any of Galaxy's internals (xref https://github.com/galaxyproject/galaxy/pull/4536). So I guess I slightly prefer the current variant - but I can make the change if you still would like to see it.

nsoranzo replied:
True, but:
- if the state is already `cancelled` at this point, the test would not fail
- it'd more difficult to debug the problem if the `state` here is `failed`, but the test fails only later.

jmchilton replied:
I'm not sure I agree - but I've made a bunch of changes including this and I like the tests as a whole better now.

jmchilton committed:
Rework invocation state tracking in API tests.

At the insistence of @nsoranzo, be more explicit about waiting for workflow test in API tests. While I was in there I also setup new abstractions for waiting on state and updated various workflow cancelling tests to use test_history context for better error reporting.

nsoranzo commented on test/api/test_workflows.py:
```diff
@@ -964,121 +964,143 @@ def test_workflow_request(self):
+            self._assert_workflow_non_terminal(uploaded_workflow_id, invocation_id)
 
-            time.sleep(.5)
+            # Review the pasued steps to allow the workflow to continue.
```
s/pasued/paused/

nsoranzo commented on test/api/test_workflows.py:
```diff
@@ -1867,6 +1889,48 @@ def test_invocations_not_accessible_by_different_user_for_published_workflow(sel
+
+        return target_state_reached
+
+    def _assert_workflow_non_terminal(self, workflow_id, invocation_id):
```
Shouldn't the function name be `_assert_invocation_non_terminal` ?

jmchilton commented:
@nsoranzo Fixed up your last two comments and rebased, thanks for the catches.

nsoranzo commented:
Thanks @jmchilton!

jmchilton commented:
Thanks for detailed review and merge @nsoranzo !
